### ISSUE TITLE
InstructionFolderPass: Crash when folding PHI nodes and erasing unused incoming values

### ISSUE THREAD

**Issue by alessandrogario** [lifting-bits/anvill#150]
Test case: [anvill_issue_150_test_case.json.zip](https://github.com/lifting-bits/anvill/files/6232772/anvill_issue_150_test_case.json.zip)

Found in commit 1ce66508c92ab146b33716f07ef5a151dce458eb

```
bool InstructionFolderPass::FoldPHINode(
    InstructionFolderPass::InstructionList &output, llvm::Instruction *instr) {
```
```
  // Finally, drop all the instructions we no longer need; the `PHINode`
  // instruction will be deleted later by the caller
  auto function_changed = !inst_replacement_list.empty();
  PerformInstructionReplacements(inst_replacement_list);

  for (auto &incoming_value : incoming_value_list) {
    auto value_as_instr =
        llvm::dyn_cast<llvm::Instruction>(incoming_value.value);

    if (value_as_instr == nullptr) {
      continue;
    }

    if (value_as_instr->use_empty()) {
      value_as_instr->eraseFromParent(); // crash here
    }
  }
```

**Comment by pgoodman** [comment]
![image](https://user-images.githubusercontent.com/43664/113068493-6459fc80-918c-11eb-9d0a-386a1c61ff27.png)

**Comment by pgoodman** [comment]
![image](https://user-images.githubusercontent.com/43664/113068633-a71bd480-918c-11eb-9961-c32fe3f29d64.png)
![image](https://user-images.githubusercontent.com/43664/113068646-b0a53c80-918c-11eb-91b4-178f07f2b83d.png)

**Comment by pgoodman** [comment]
Possible fix:

```c++
  std::unordered_set<llvm::Instruction *> to_remove;
  for (auto &incoming_value : incoming_value_list) {
    auto value_as_instr =
        llvm::dyn_cast<llvm::Instruction>(incoming_value.value);
    if (value_as_instr && value_as_instr->use_empty()) {
      to_remove.insert(value_as_instr);
    }
  }
  for (auto value_as_instr : to_remove) {
    value_as_instr->eraseFromParent();
    function_changed = true;
  }
```

**Comment by alessandrogario** [comment]
Fixed in 0fce7c9a00f464428e58858d6d3fd78c7597bdff
